source/common/config/config_provider_impl.h

```diff
@@ -0,0 +1,364 @@
+
+/**
+ * Provides generic functionality required by all dynamic config providers, including distribution
+ * of config updates to all workers.
+ *
+ * This class can not be instantiated directly; instead, it provides the foundation for
+ * dynamic config provider implementations which derive from it.
+ */
+class DynamicConfigProviderImpl : public ConfigProvider {
+public:
+  DynamicConfigProviderImpl() = delete;
+
+  virtual ~DynamicConfigProviderImpl() { subscription_->unbindConfigProvider(this); }
+
+  // Envoy::Config::ConfigProvider
+  SystemTime lastUpdated() const override { return subscription_->lastUpdated(); }
+
+  // Envoy::Config::ConfigProvider
+  ConfigConstSharedPtr getConfig() const override {
+    return tls_->getTyped<ThreadLocalConfig>().config_;
+  }
+
+  virtual ConfigConstSharedPtr onConfigProtoUpdate(const Protobuf::Message& config_proto) PURE;
+
+  /**
+   * Must be called by the derived class' constructor.
+   * @param initial_config supplies an initial Envoy::Config::ConfigProvider::Config associated with
+   *                       the underlying subscription.
+   */
+  void initialize(ConfigConstSharedPtr initial_config) {
+    subscription_->bindConfigProvider(this);
+    tls_->set([initial_config](Event::Dispatcher&) -> ThreadLocal::ThreadLocalObjectSharedPtr {
+      return std::make_shared<ThreadLocalConfig>(initial_config);
+    });
+  }
+
+  /**
+   * Propagates a newly instantiated Envoy::Config::ConfigProvider::Config to all workers.
+   * @param config supplies the newly instantiated config.
+   */
+  void onConfigUpdate(ConfigConstSharedPtr config) {
+    tls_->runOnAllThreads(
+        [this, config]() -> void { tls_->getTyped<ThreadLocalConfig>().config_ = config; });
+  }
+
+protected:
+  DynamicConfigProviderImpl(ConfigSubscriptionInstanceSharedPtr&& subscription,
+                            Server::Configuration::FactoryContext& factory_context)
+      : subscription_(subscription), tls_(factory_context.threadLocal().allocateSlot()) {}
+
+  const ConfigSubscriptionInstanceSharedPtr& subscription() const { return subscription_; }
+
+private:
+  struct ThreadLocalConfig : public ThreadLocal::ThreadLocalObject {
+    ThreadLocalConfig(ConfigProvider::ConfigConstSharedPtr initial_config)
+        : config_(initial_config) {}
+
+    ConfigProvider::ConfigConstSharedPtr config_;
+  };
+
+  ConfigSubscriptionInstanceSharedPtr subscription_;
+  ThreadLocal::SlotPtr tls_;
+};
+
+/**
+ * Provides generic functionality required by all config provider managers, such as managing shared
+ * lifetime of subscriptions and dynamic config providers, along with determining which
+ * subscriptions should be associated with newly instantiated providers.
+ *
+ * This class can not be instantiated directly; instead, it provides the foundation for
+ * dynamic config provider implementations which derive from it.
+ */
+class ConfigProviderManagerImpl : public ConfigProviderManager, public Singleton::Instance {
```

Comment:
Name it `ConfigProviderManagerImplBase` if it cannot instantiate directly.

Reply:
Done. Renamed all base classes to append Base suffix.